### PR TITLE
cpu_device_hotpluggable: Add a workaround for win2016 guest

### PR DESCRIPTION
**Pull request by peixiu**
For win2016 guest, there is a exist microsoft issue:
https://support.huawei.com/enterprise/zh/doc/EDOC1100034211/5ba99a60
In order to workaround the known issue, add install driver of
"HID Button over Interrupt Driver" steps before cpu hotplugged.

id: 1822451
Signed-off-by: Peixiu Hou <phou@redhat.com>

### CODE REVIEW

**Comment by peixiu** [comment]
Before fix, the case run on win2016 will hit "CPU quantity mismatched!" error, this fix solve this problem.
But due to another Microsoft issue(https://bugzilla.redhat.com/show_bug.cgi?id=1445603#c20), will cause the win2016 guest vm reset after a cpu hot-plugged, that will cause these jobs  with_reboot.shell_reboot/with_shutdown.shell_shutdown/with_migrate failed, it's not related this patch fix.

Test results: 
 (1/6) type_specific.io-github-autotest-qemu.cpu_device_hotpluggable.hotplug.single_vcpu.with_reboot.shell_reboot: ERROR: Guest refuses to go down (462.90 s)
 (2/6) type_specific.io-github-autotest-qemu.cpu_device_hotpluggable.hotplug.single_vcpu.with_reboot.qemu_system_reset: PASS (121.01 s)
 (3/6) type_specific.io-github-autotest-qemu.cpu_device_hotpluggable.hotplug.single_vcpu.with_shutdown.shell_shutdown: FAIL: Guest refuses to go down after vcpu hotplug (430.32 s)
 (4/6) type_specific.io-github-autotest-qemu.cpu_device_hotpluggable.hotplug.single_vcpu.with_shutdown.qemu_system_powerdown:PASS (180.10 s)
 (5/6) type_specific.io-github-autotest-qemu.cpu_device_hotpluggable.hotplug.single_vcpu.with_migrate: ERROR: Guest refuses to go down (520.70 s)
 (6/6) type_specific.io-github-autotest-qemu.cpu_device_hotpluggable.hotplug.single_vcpu.with_pause_resume: PASS (227.84 s)

**Comment by peixiu** [comment]
Hi @huangyum @PaulYuuu, Could you help to review this patch?

Thanks a lot~
Peixiu

**Comment by PaulYuuu** [comment]
Is it more suitable as a variant of `single_driver_install`?

**Comment by peixiu** [comment]
> Is it more suitable as a variant of `single_driver_install`?

Hi yihuang, I also considering it before, I set to this file because follow reasons:
1. single_driver_install is for our virtio-win driver product testing, it's covered all virtio-win single driver(balloon, viorng, netkvm, vioscsi, viostor, vioinput, vioser) install, but "HID Button over Interrupt Driver" doesn't included in virtio-win driver.
2. "HID Button over Interrupt Driver" only shown on win2016 guest vm, it will occupy hot-added cpu's driver, cause the system cannot detected new added cpu, it's strong related with cpu hotplug.
As above, I thinks set this function to cpu_device_hotpluggable is more suitable.

Thanks~
Peixiu

File: qemu/tests/cpu_device_hotpluggable.py [hunk 1]
```diff
@@ -132,10 +133,36 @@ def sub_pause_resume():
         pluggable_vcpu = -(vcpus_count * len(pluggable_vcpu_dev))
     expected_vcpus = vm.get_cpu_count() + pluggable_vcpu
 
+    # Add a workaround for win2016 guest to solve the issue that
+    # https://support.huawei.com/enterprise/zh/doc/EDOC1100034211/5ba99a60.
+    # Add steps for install driver of "HID Button over Interrupt Driver".
```
Comment (huangyum) [inline qemu/tests/cpu_device_hotpluggable.py:138]: There are other cpu hotplug test cases need this workaround. Suggest to wrap it in a single function and invoke it in related test cases.

Comment (peixiu) [reply qemu/tests/cpu_device_hotpluggable.py:138]: Hi yuhuang, I checked the privider directory, only cpu_utils.py related this case, but it doesn't suitable setting this function in. It's also as a case level setting, seems not suitable set in avocado vt as well, How do you think create a new file for this workaround under provider directory? thus convenient other cpu hotplug test cases to call it and also can avoid case call case. cc @PaulYuuu

Comment (huangyum) [reply qemu/tests/cpu_device_hotpluggable.py:138]: Sounds good to me.

Comment (peixiu) [reply qemu/tests/cpu_device_hotpluggable.py:138]: Or do you have any better advice, thanks  a lot~@huangyum @PaulYuuu

**Comment by PaulYuuu** [comment]
> > Is it more suitable as a variant of `single_driver_install`?
> 
> Hi yihuang, I also considering it before, I set to this file because follow reasons:
> 
> 1. single_driver_install is for our virtio-win driver product testing, it's covered all virtio-win single driver(balloon, viorng, netkvm, vioscsi, viostor, vioinput, vioser) install, but "HID Button over Interrupt Driver" doesn't included in virtio-win driver.
> 2. "HID Button over Interrupt Driver" only shown on win2016 guest vm, it will occupy hot-added cpu's driver, cause the system cannot detected new added cpu, it's strong related with cpu hotplug.
>    As above, I thinks set this function to cpu_device_hotpluggable is more suitable.
> 
> Thanks~
> Peixiu

Just consider that this is the problem of Win2016 itself, not the case step is wrong, so it is reasonable to prepare a good guest before running cpu hotplug cases.

But I usually don’t test windows guests, so as long as you think it’s okay, I’m acceptable.

**Comment by peixiu** [comment]
> > > Is it more suitable as a variant of `single_driver_install`?
> > 
> > 
> > Hi yihuang, I also considering it before, I set to this file because follow reasons:
> > 
> > 1. single_driver_install is for our virtio-win driver product testing, it's covered all virtio-win single driver(balloon, viorng, netkvm, vioscsi, viostor, vioinput, vioser) install, but "HID Button over Interrupt Driver" doesn't included in virtio-win driver.
> > 2. "HID Button over Interrupt Driver" only shown on win2016 guest vm, it will occupy hot-added cpu's driver, cause the system cannot detected new added cpu, it's strong related with cpu hotplug.
> >    As above, I thinks set this function to cpu_device_hotpluggable is more suitable.
> > 
> > Thanks~
> > Peixiu
> 
> Just consider that this is the problem of Win2016 itself, not the case step is wrong, so it is reasonable to prepare a good guest before running cpu hotplug cases.
> 
> But I usually don’t test windows guests, so as long as you think it’s okay, I’m acceptable.

OK, thanks, will consider your advise~

**Comment by peixiu** [comment]
Tested with 1 job:
PASS 1-Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.cpu_device_hotpluggable.hotplug.single_vcpu.with_reboot.qemu_system_reset

File: qemu/tests/cpu_device_hotpluggable.py [hunk 2]
```diff
@@ -134,10 +135,14 @@ def sub_pause_resume():
     else:
         pluggable_vcpu_dev = vcpu_devices[::-1]
 
+    if params.get("workaround_need", '') == "yes":
```
Comment (PaulYuuu) [inline qemu/tests/cpu_device_hotpluggable.py:138]: `params.get_boolean("workaround_need")`

File: qemu/tests/cpu_device_hotpluggable.py (hunk 1 above)
Comment (peixiu) [reply qemu/tests/cpu_device_hotpluggable.py]: Updated, thanks~

File: qemu/tests/cpu_device_hotpluggable.py (hunk 2 above)
Comment (peixiu) [reply qemu/tests/cpu_device_hotpluggable.py]: Updated, thanks~

**Comment by peixiu** [comment]
Test result with latest code:
PASS 1-Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.cpu_device_hotpluggable.hotplug.single_vcpu.with_reboot.qemu_system_reset

@PaulYuuu @huangyum I updated the code,  could you help to review it ? 
thanks a lot~

File: provider/win_wora.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    cause the system cannot detected new added cpu, need modify the driver.
+    issue details please refer to:
+    https://support.huawei.com/enterprise/zh/doc/EDOC1100034211/5ba99a60.
+    """
+    devcon_path = utils_misc.set_winutils_letter(session,
+                                                 params["devcon_path"])
+    dev_hwid = params["dev_hwid"]
+    chk_cmd = '%s find %s' % (devcon_path, dev_hwid)
+    chk_pat = r'ACPI\\ACPI0010.*\: Generic Bus'
+    while not re.search(chk_pat, session.cmd(chk_cmd)):
```
Comment (PaulYuuu) [inline provider/win_wora.py:26]: Is there any reason to use `while`?

Comment (peixiu) [reply provider/win_wora.py:26]: Want to run twice chk_cmd command here, first time, check if the correct driver is installed, if not, run install command; second time, double check if the correct driver is installed.

Comment (PaulYuuu) [reply provider/win_wora.py:26]: Okay, thanks. Only one note: if the driver installation failed, this `while` loop will never terminate.

**Review by PaulYuuu — APPROVED** [review]

**Comment by huangyum** [comment]
Would you please update other cpu hotplug test cases as well?  Such as cpu_device_hotpluggable_with_numa.py, cpu_device_hotpluggable_with_stress.py, cpu_device_hotplug_maximum.py.

File: provider/win_wora.py (hunk 3 above)
Comment (huangyum) [reply provider/win_wora.py:26]: Prefer to raise error if installation failed instead of keeping running endless loop.

**Comment by peixiu** [comment]
> Would you please update other cpu hotplug test cases as well? Such as cpu_device_hotpluggable_with_numa.py, cpu_device_hotpluggable_with_stress.py, cpu_device_hotplug_maximum.py.

Do you mean all follows cases?
cpu_device_hotplug_during_boot.py       
cpu_device_hotpluggable_with_numa.py    
cpu_device_hotplug_maximum.py           
cpu_device_hotpluggable_with_stress.py 
cpu_device_hotplug.py

**Comment by peixiu** [comment]
> Would you please update other cpu hotplug test cases as well? Such as cpu_device_hotpluggable_with_numa.py, cpu_device_hotpluggable_with_stress.py, cpu_device_hotplug_maximum.py.

It seems they are not support win2016 guest, I try to list these case jobs, but reply no test references provided, could you help to confirm it? thanks~

python3 ConfigTest.py --testcase=cpu_device_hotplug_during_boot,cpu_device_hotplug,cpu_device_hotpluggable_with_numa,cpu_device_hotpluggable_with_stresscpu_device_hotplug_maximum --guestname=Win2016 --platform=x86_64 --clone=no --driveformat=virtio_scsi  --dryrun

**Test Suite could not be create. No test references provided nor any other arguments resolved into tests**

File: provider/win_wora.py (hunk 3 above)
Comment (peixiu) [reply provider/win_wora.py]: Updated, thanks~

**Comment by huangyum** [comment]
> > Would you please update other cpu hotplug test cases as well? Such as cpu_device_hotpluggable_with_numa.py, cpu_device_hotpluggable_with_stress.py, cpu_device_hotplug_maximum.py.
> 
> It seems they are not support win2016 guest, I try to list these case jobs, but reply no test references provided, could you help to confirm it? thanks~
> 
> python3 ConfigTest.py --testcase=cpu_device_hotplug_during_boot,cpu_device_hotplug,cpu_device_hotpluggable_with_numa,cpu_device_hotpluggable_with_stresscpu_device_hotplug_maximum --guestname=Win2016 --platform=x86_64 --clone=no --driveformat=virtio_scsi --dryrun
> 
> **Test Suite could not be create. No test references provided nor any other arguments resolved into tests**

They are scripts name, not case name. 

The related cfg and py are as below:

cpu_device_hotpluggable.cfg
cpu_device_hotplug_maximum.cfg
cpu_device_hotpluggable.py
cpu_device_hotpluggable_with_numa.py
cpu_device_hotpluggable_with_stress.py
cpu_device_hotplug_maximum.py

**Comment by peixiu** [comment]
@huangyum Please help to review this patch again, I update the workaround related code on other cpu_device_hotplug* files as your mentioned, Thanks a lot~

Peixiu

**Comment by huangyum** [comment]
There is one more case need to update, "invalid_cpu_device_hotplug".

**Comment by peixiu** [comment]
> There is one more case need to update, "invalid_cpu_device_hotplug".

@huangyum Ok, updated the invalid_cpu_device_hotplug.py and invalid_cpu_device_hotplug.cfg, please help to review again, thanks :)~

**Review by huangyum — APPROVED** [review]
LGTM

**Pull request merged**